src/vsr/grid.zig

```diff
@@ -282,32 +296,64 @@ pub fn GridType(comptime Storage: type) type {
+            assert(trailer.callback == .none);
+            const grid: *Grid = @fieldParentPtr("free_set_checkpoint_blocks_acquired", trailer);
+            grid.open_free_set_callback();
+        }
+
+        fn open_free_set_callback_blocks_released(trailer: *CheckpointTrailer) void {
+            assert(trailer.callback == .none);
+            const grid: *Grid = @fieldParentPtr("free_set_checkpoint_blocks_released", trailer);
+            grid.open_free_set_callback();
+        }
+
+        fn open_free_set_callback(grid: *Grid) void {
             const callback = grid.callback.open;
+            // May still be reading the CheckpointTrailer for `blocks_acquired`.
+            if (grid.free_set_checkpoint_blocks_acquired.callback == .open) return;
+            assert(grid.free_set_checkpoint_blocks_acquired.callback == .none);
+
+            // May still be reading the CheckpointTrailer for `blocks_released`.
+            if (grid.free_set_checkpoint_blocks_released.callback == .open) return;
+            assert(grid.free_set_checkpoint_blocks_released.callback == .none);
 
             {
                 assert(!grid.free_set.opened);
                 defer assert(grid.free_set.opened);
 
-                const free_set_checkpoint_block_addresses =
-                    free_set_checkpoint.block_addresses[0..free_set_checkpoint.block_count()];
+                const block_count_encoded_blocks_acquired =
+                    grid.free_set_checkpoint_blocks_acquired.block_count();
+                const block_count_encoded_blocks_released =
+                    grid.free_set_checkpoint_blocks_released.block_count();
                 grid.free_set.open(.{
-                    .encoded = free_set_checkpoint.decode_chunks(),
-                    .block_addresses = free_set_checkpoint_block_addresses,
+                    .encoded = .{
+                        .blocks_acquired = grid.free_set_checkpoint_blocks_acquired.decode_chunks(),
+                        .blocks_released = grid.free_set_checkpoint_blocks_released.decode_chunks(),
+                    },
+                    .block_addresses = .{
+                        .blocks_acquired = grid.free_set_checkpoint_blocks_acquired
+                            .block_addresses[0..block_count_encoded_blocks_acquired],
+                        .blocks_released = grid.free_set_checkpoint_blocks_released
+                            .block_addresses[0..block_count_encoded_blocks_released],
+                    },
                 });
-                assert((grid.free_set.count_acquired() > 0) == (free_set_checkpoint.size > 0));
+                assert((grid.free_set.count_acquired() > 0) ==
+                    (grid.free_set_checkpoint_blocks_acquired.size > 0));
                 assert(grid.free_set.count_reservations() == 0);
-                assert(grid.free_set.count_released() == grid.free_set_checkpoint.block_count());
```

Comment:
We can still assert:
```zig
assert(grid.free_set.count_released() >= grid.free_set_checkpoint.block_count());
```
(Note that `>=` rather than `==`.)